File: libcomponent/src/main/java/com/bear/libcomponent/component/ComponentFragment.java
package com.bear.libcomponent.component;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewbinding.ViewBinding;

import com.bear.libbase.fragment.BaseFragment;

public abstract class ComponentFragment<VB extends ViewBinding> extends BaseFragment<VB> {
    
    @Override
    @CallSuper
    public void onAttach(Context context) {
        super.onAttach(context);
        getComponentManager().dispatchOnAttach(this, context);
    }

    @Nullable
    @Override
    @CallSuper
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View contentView = super.onCreateView(inflater, container, savedInstanceState);
        getComponentManager().dispatchOnCreateView(this, getBinding());
        return contentView;
    }

    @Override
    @CallSuper
    public void onDestroyView() {
        super.onDestroyView();
        getComponentManager().dispatchOnDestroyView(this);
    }

    @Override
    @CallSuper
    public void onDetach() {
        super.onDetach();
        getComponentManager().dispatchOnDetach(this);
    }

    @Override
    protected void onFirstVisible() {
        getComponentManager().dispatchOnFirstVisible(this);
    }

    public void regFragComponent(FragmentComponent component, Object tag) {
        getComponentManager().regComponent(this, component, tag);
    }

    public void regFragComponent(FragmentComponent component) {
        getComponentManager().regComponent(this, component);
    }

    public <C extends GroupComponent> C getComponent(Class<C> clz, Object tag) {
        return getComponentManager().getComponent(clz, tag);
    }

    public <C extends GroupComponent> C getComponent(Class<C> clz) {
        return getComponentManager().getComponent(clz);
    }
    
    private ComponentManager getComponentManager() {
        if (getActivity() instanceof ComponentActivity) {
            return ((ComponentActivity) getActivity()).getComponentManager();
        }
        throw new RuntimeException("getComponentManager return null");
    }

    @Override
    protected abstract VB inflateViewBinding(LayoutInflater inflater, ViewGroup container);
}

File: libcomponent/src/main/java/com/bear/libcomponent/host/ComponentFragment.kt
package com.bear.libcomponent.host

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.viewbinding.ViewBinding
import com.bear.libbase.fragment.BaseFragment
import com.bear.libcomponent.component.base.GroupComponent
import com.bear.libcomponent.component.ui.FragmentComponent
import com.bear.libcomponent.component.ui.NonUIComponent
import com.bear.libcomponent.component.ui.ViewComponent
import com.bear.libcomponent.core.ComponentManager

abstract class ComponentFragment<VB : ViewBinding> : BaseFragment<VB>() {
    private val componentManager: ComponentManager
        get() {
            if (activity is ComponentActivity<*>) {
                return (activity as ComponentActivity<*>).componentManager
            }
            throw RuntimeException("getComponentManager return null")
        }

    @CallSuper
    override fun onAttach(context: Context) {
        super.onAttach(context)
        componentManager.dispatchOnAttach(this, context)
    }

    @CallSuper
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val contentView = super.onCreateView(inflater, container, savedInstanceState)
        componentManager.dispatchOnCreateView(this, binding)
        return contentView
    }

    @CallSuper
    override fun onDestroyView() {
        super.onDestroyView()
        componentManager.dispatchOnDestroyView(this)
    }

    @CallSuper
    override fun onDetach() {
        super.onDetach()
        componentManager.dispatchOnDetach(this)
    }

    override fun onFirstVisible() {
        componentManager.dispatchOnFirstVisible(this)
    }

    @JvmOverloads
    fun regComponent(component: FragmentComponent<VB>, tag: Any? = null) {
        componentManager.regComponent(this, component, tag)
    }

    @JvmOverloads
    fun regComponent(component: ViewComponent<*>, tag: Any? = null) {
        componentManager.regComponent(requireContext(), component, tag)
    }

    @JvmOverloads
    fun regComponent(component: NonUIComponent, tag: Any? = null) {
        componentManager.regComponent(requireContext(), component, tag)
    }

    @JvmOverloads
    fun <C : GroupComponent> getComponent(clz: Class<C>, tag: Any? = null): C? {
        return componentManager.getComponent(clz, tag)
    }

    abstract override fun inflateViewBinding(inflater: LayoutInflater, container: ViewGroup?): VB
}
